cruise-control/src/main/scala/com/linkedin/kafka/cruisecontrol/executor/ExecutorUtils.scala

```diff
@@ -85,38 +87,76 @@ object ExecutorUtils {
+                    addTask = false
+                    Seq.empty
                   } else {
-                    // If the source broker is no longer in the list, just do not add the task to the reassignment.
+                    // Get new replicas for replica addition.
+                    getNewReplicasForReplicaAddition(currentReplicaAssignment, task)
+                  }
+                } else if (task.proposal.balancingAction() == BalancingAction.REPLICA_DELETION) {
+                  if (!currentReplicaAssignment.contains(sourceBroker)) {
+                    // Reassignment is done already.
                     addTask = false
-                    currentReplicaAssignment
+                    Seq.empty
+                  } else {
+                    // Get new replicas for replica deletion.
+                    getNewReplicasForReplicaDeletion(currentReplicaAssignment, task)
                   }
+                } else {
+                  throw new RuntimeException(s"Unexpected balancing action for proposal ${task.proposal} for [$topic, $partition]")
                 }
               }
             }
         }
         if (addTask)
-          newPartitionAssignment += (tp -> newReplicas)
+          newReplicaAssignment += (tp -> newReplicas)
       })
 
       // We do not use the ReassignPartitionsCommand here because we want to have incremental partition movement.
-      if (newPartitionAssignment.nonEmpty)
-        zkUtils.updatePartitionReassignmentData(newPartitionAssignment)
+      if (newReplicaAssignment.nonEmpty)
+        zkUtils.updatePartitionReassignmentData(newReplicaAssignment)
+    }
+  }
+
+  def getNewReplicasForReplicaAddition(currentReplicaAssignment: Seq[Int], task: ExecutionTask): Seq[Int] = {
+    // Source broker is guaranteed to be a null for replica addition.
+    val destinationBroker = task.destinationBrokerId()
+    currentReplicaAssignment :+ destinationBroker.toInt
+  }
+
+  def getNewReplicasForReplicaDeletion(currentReplicaAssignment: Seq[Int], task: ExecutionTask): Seq[Int] = {
+    // Destination broker is guaranteed to be a null for replica deletion.
+    val sourceBroker = task.sourceBrokerId()
+    currentReplicaAssignment.filter(_ != sourceBroker)
+  }
+
+  def getNewReplicasForReplicaMovement(currentReplicaAssignment: Seq[Int], task: ExecutionTask): Seq[Int] = {
+    // Both source and destination brokers are guaranteed to be non-null for replica movement.
+    val sourceBroker = task.sourceBrokerId()
+    val destinationBroker = task.destinationBrokerId()
+    if (!currentReplicaAssignment.contains(destinationBroker)) {
+      if (currentReplicaAssignment.contains(sourceBroker)) {
+        // This is a normal movement.
+        (currentReplicaAssignment :+ destinationBroker.toInt).filter(_ != sourceBroker)
+      } else {
+        throw new IllegalStateException(s"Unexpected proposal ${task.proposal}. Neither source nor destination broker" +
```

Comment:
Minior: we may want to include the current replica assignment in the exception message as well.